{"metadata":{"kernelspec":{"display_name":"Bash","language":"bash","name":"bash"},"language_info":{"codemirror_mode":"shell","file_extension":".sh","mimetype":"text/x-sh","name":"bash"}},"nbformat_minor":4,"nbformat":4,"cells":[{"cell_type":"markdown","source":"## Aligning Reads to a Reference Genome with BWA\n### Bioinformatics Coffee Chat - April 7, 2020\n### Programs used in this session\n#### [BWA](http://bio-bwa.sourceforge.net/), [Picard](https://broadinstitute.github.io/picard/)","metadata":{}},{"cell_type":"markdown","source":"Let's get started!\n\nWe have two directories:\n* `00_genome` - reference sequence\n* `01_fastqs` - raw Illumina reads to map to the genome","metadata":{}},{"cell_type":"markdown","source":"#### Fastq -> SAM -> BAM\n\n#### Process reference genome\nBWA requires building an index for your reference genome to allow it to more efficiently search the genome during sequence alignment:","metadata":{}},{"cell_type":"code","source":"bwa index -p 00_genome/Falb 00_genome/Falbicolis.chr5.fa.gz","metadata":{"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"You should have several new files in the `00_genome` directory that all start with 'Falb', since this is the value we gave after the `-p` flag.\n\nLets use BWA to align our reads to the reference genome and get a SAM file (Sequence Alignment/Map), one for each sample. For information about BWA, simply type its name, followed by `--help`. This procedure works for most programs.","metadata":{}},{"cell_type":"code","source":"mkdir -p 02_bams\nfor INDEX in 1 2 3 34 35; \ndo\n   bwa mem -M -t 1 -R \"@RG\\tID:COL_${INDEX}\\tSM:COL_${INDEX}\" 00_genome/Falb \\\n   01_fastqs/Falb_COL${INDEX}.1.fastq.gz \\\n   01_fastqs/Falb_COL${INDEX}.2.fastq.gz \\\n   > 02_bams/Falb_COL${INDEX}.sam #2> Falb_COL${INDEX}.log\ndone","metadata":{"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"Note: read groups refer to sets of reads that were generated from a single run of a sequencing instrument, and this information is stored in our SAM file on lines that start with `@RG`.\nUsing read groups allows us to not just distinguish between samples, but also particular samples that were sequenced across several experiments.\nPrograms like the GATK require this information so that it can attempt to compensate for variability between sequencing runs.\n\nNext, lets convert our SAM files from BWA to BAM files, which are compressed versions that a lot of downstream programs use as input files.","metadata":{}},{"cell_type":"code","source":"for INDEX in 1 2 3 34 35;\ndo\n  picard SortSam \\\n  I=02_bams/Falb_COL${INDEX}.sam \\\n  O=02_bams/Falb_COL${INDEX}.sorted.bam \\\n  SORT_ORDER=coordinate \\\n  CREATE_INDEX=true\ndone","metadata":{"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"Last, we need to create an index of our BAM file in order for downstream programs to quickly access its contents.","metadata":{}},{"cell_type":"code","source":"for INDEX in 1 2 3 34 35\ndo\n  picard BuildBamIndex \\\n  I=02_bams/Falb_COL${INDEX}.sorted.bam\ndone","metadata":{"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"### That's all, folks!","metadata":{}}]}